File: verification/julia_levels/l4_nse_2d.jl
# l4_nse_2d.jl — 2D NSE: тождества поворота + энергетика (собственный FFT radix-2)
include(joinpath(@__DIR__, "common.jl"))

const COSB = sqrt(1 - B^2)
const SINB = B

"""FFT2 (по обеим осям) для матрицы n×n через myfft (последняя ось + перестановка)."""
function fft2(A::Matrix{ComplexF64}; inverse::Bool = false)
    T = myfft(A; inverse)
    return Matrix{ComplexF64}(permutedims(myfft(permutedims(T, (2, 1)); inverse), (2, 1)))
end

function run_l4(; N = 64, nu = 0.001, T = 2.0, dt = 0.002)
    selftest = fft_selftest(64)
    crit = N ÷ 3
    NP = 128                     # степень двойки: 2N-паддинг — произведения без алиасинга
    k = fftfreq_r2c(N)           # целые волновые числа
    kP = fftfreq_r2c(NP)
    # сетки: kx по строкам (полная ось), ky по столбцам (rfft-половина не нужна — полный комплексный формат)
    KX = [k[j] for i in 1:N, j in 1:N]
    KY = [k[i] for i in 1:N, j in 1:N]
    K2 = KX.^2 .+ KY.^2; K2[1, 1] = 1.0
    KXP = [kP[j] for i in 1:NP, j in 1:NP]
    KYP = [kP[i] for i in 1:NP, j in 1:NP]
    K2P = KXP.^2 .+ KYP.^2; K2P[1, 1] = 1.0
    x = range(0; length = N, step = 2pi / N)
    X = [Float64(x[j]) for i in 1:N, j in 1:N]
    Y = [Float64(x[i]) for i in 1:N, j in 1:N]

    mask = (abs.(KX) .<= crit) .& (abs.(KY) .<= crit)
    proj(f) = begin
        p = (KX .* f[1] .+ KY .* f[2]) ./ K2
        [f[1] .- KX .* p, f[2] .- KY .* p]
    end
    to_spec(f) = fft2(ComplexF64.(f))
    to_real(f) = real.(fft2(f; inverse = true))

    u0 = sin.(X) .* cos.(Y)
    v0 = -cos.(X) .* sin.(Y)
    wh = proj(to_spec([u0, v0]))
    om0 = to_real([1im .* KY .* wh[2] .- 1im .* KX .* wh[0 + 1]])[1]

    # (i) тождество ω' = cos θ_b·ω на 200 div-free полях
    rng = MersenneTwister(7)
    max_id = 0.0
    for _ in 1:200
        f = proj(to_spec([randn(rng, N, N), randn(rng, N, N)]))
        om = to_real([1im .* KY .* f[2] .- 1im .* KX .* f[1]])[1]
        rot = [COSB .* f[1] .- SINB .* f[2], SINB .* f[1] .+ COSB .* f[2]]
        omr = to_real([1im .* KY .* rot[2] .- 1im .* KX .* rot[1]])[1]
        max_id = max(max_id, maximum(abs.(omr .- COSB .* om)))
    end
    w = [u0, v0]
    E = sum(abs2, w[1]) + sum(abs2, w[2])
    wrot = [COSB .* w[1] .- SINB .* w[2], SINB .* w[1] .+ COSB .* w[2]]
    energy_err = abs((sum(abs2, wrot[1]) + sum(abs2, wrot[2])) - E) / E

    # честная реализация rhs (вектор из 2 спектральных полей)
    function rhs2(uh)
        uh2 = [embed(f, crit, NP) for f in uh]
        u = [real.(fft2(f; inverse = true)) for f in uh2]
        dX = [real.(fft2(1im .* KXP .* f; inverse = true)) for f in uh2]
        dY = [real.(fft2(1im .* KYP .* f; inverse = true)) for f in uh2]
        adv = [u[1] .* dX[1] .+ u[2] .* dY[1], u[1] .* dX[2] .+ u[2] .* dY[2]]
        nlp = to_spec(adv)
        p = (KXP .* nlp[1] .+ KYP .* nlp[2]) ./ K2P
        nlp = [nlp[1] .- KXP .* p, nlp[2] .- KYP .* p]
        out = [cut(nlp[1], crit, N), cut(nlp[2], crit, N)]
        return [.-out[1] .- nu .* K2 .* uh[1], .-out[2] .- nu .* K2 .* uh[2]]
    end

    n_steps = Int(round(T / dt))
    om_max0 = maximum(abs.(om0))
    om_viol = 0.0
    phi = dt * THETA
    cph, sph = cos(phi), sin(phi)
    max_dE = 0.0
    E_prev = (sum(abs2, u0) + sum(abs2, v0)) / 2 / N^2
    diss = 0.0; Z_prev = nothing
    wh_run = wh
    for n in 1:n_steps
        k1 = rhs2(wh_run)
        k2 = rhs2(wh_run .+ 0.5dt .* k1)
        k3 = rhs2(wh_run .+ 0.5dt .* k2)
        k4 = rhs2(wh_run .+ dt .* k3)
        wh_run = wh_run .+ (dt / 6) .* (k1 .+ 2k2 .+ 2k3 .+ k4)
        om = to_real([1im .* KY .* wh_run[2] .- 1im .* KX .* wh_run[1]])[1]
        om_viol = max(om_viol, maximum(abs.(om)) - om_max0)
        Z = 0.5 * sum(abs2, om) / N^2
        if Z_prev !== nothing
            diss += nu * dt * (Z_prev + Z)
        end
        Z_prev = Z
        uu = to_real(wh_run)
        E_pre = (sum(abs2, uu[1]) + sum(abs2, uu[2])) / 2 / N^2
        wh_rot = proj([cph .* wh_run[1] .- sph .* wh_run[2], sph .* wh_run[1] .+ cph .* wh_run[2]])
        uu2 = to_real(wh_rot)
        E_post = (sum(abs2, uu2[1]) + sum(abs2, uu2[2])) / 2 / N^2
        max_dE = max(max_dE, abs(E_post - E_pre))
        wh_run = wh_rot
    end
    balance = abs((sum(abs2, to_real(wh_run)[1]) + sum(abs2, to_real(wh_run)[2])) / 2 / N^2 - E_prev + diss)

    ok = selftest < 1e-10 && max_id < 1e-10 && energy_err < 1e-12 && om_viol < 1e-9 &&
         max_dE < 1e-7 && balance < 1e-3
    out = Dict("level" => "L4",
        "params" => Dict("N" => N, "nu" => nu, "T" => T, "dt" => dt,
                         "theta_b_deg" => float(rad2deg(THETA)),
                         "protocol" => "continuous dt·θ_b; 2/3-правило; 2N-паддинг; собственный FFT"),
        "checks" => Dict(
            "FFT самотест против DFT" => selftest,
            "тождество ω' = cos θ_b·ω" => max_id,
            "энергия при повороте |E'−E|/E" => energy_err,
            "максимум-принцип ω (превышение)" => om_viol,
            "|ΔE| за один поворот" => max_dE,
            "энергетический баланс" => balance,
        ), "ok" => ok)
    write_json(joinpath(RESULTS_DIR, "l4_nse_2d.json"), out)
    println("L4: ", ok ? "PASS" : "FAIL", " | FFT:", selftest, " ω'-тожд.: ", max_id)
    return ok
end

# rfftfreq-аналог для степени двойки (целые волновые числа)
function fftfreq_r2c(n::Int)
    return [j - n * (j > n ÷ 2) for j in 0:n-1]
end

# врезка спектра n-сетки (моды ≤ crit) в np-сетку и обратно (полный комплексный формат)
function embed(f, crit, np)
    out = zeros(ComplexF64, np, np)
    out[1:crit+1, 1:crit+1] = f[1:crit+1, 1:crit+1]
    out[end-crit+1:end, 1:crit+1] = f[end-crit+1:end, 1:crit+1]
    out[1:crit+1, end-crit+1:end] = f[1:crit+1, end-crit+1:end]
    out[end-crit+1:end, end-crit+1:end] = f[end-crit+1:end, end-crit+1:end]
    return out
end
function cut(f, crit, n)
    out = zeros(ComplexF64, n, n)
    out[1:crit+1, 1:crit+1] = f[1:crit+1, 1:crit+1]
    out[end-crit+1:end, 1:crit+1] = f[end-crit+1:end, 1:crit+1]
    out[1:crit+1, end-crit+1:end] = f[1:crit+1, end-crit+1:end]
    out[end-crit+1:end, end-crit+1:end] = f[end-crit+1:end, end-crit+1:end]
    return out
end

File: verification/python_levels/l4_nse_2d.py
# -*- coding: utf-8 -*-
"""
L4 — 2D NSE: точные тождества b-поворота и энергетика.

Математическое ядро (проверяется численно с машинной точностью):
  Для дивергентно-свободного u в 2D поворот u' = cos θ_b·u + b·(ẑ×u) (эквивалент
  точной формы u' = u∥ + √(1−b²)·u⊥ + b·(ω̂×u⊥), т.к. в 2D u∥ = 0, ω̂ = ẑ):
    (i)   ω' = cos θ_b · ω   — ТОЧНО (слагаемое b·(ẑ×u) даёт curl = ∇·u = 0);
    (ii)  |u'|² = |u|²        — ТОЧНО (u ⊥ ẑ×u, |ẑ×u| = |u|) — инжекции энергии нет;
    (iii) div u' = −b·ω       — градиентная часть уходит в давление (проекция Лерэ).
Численный эталон: 2D NSE (вихрь Тейлора–Грина), спектральный метод, RK4:
  * энергетическое тождество dE/dt = −2νZ (Z — энстрофия) на сетке;
  * максимум-принцип для ω;
  * непрерывный b-поворот за шаг dt·θ_b: |ΔE| на повороте ~ машинный ноль.
"""
import json, math, sys
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).parent))
from config import float_theta_b_deg

RESULTS_DIR = Path(__file__).parent / "results"
B = 1 / (4 * math.pi + 2 * math.sqrt(3))
THETA = math.asin(B)
COS, SIN = math.sqrt(1 - B * B), B


def run(N=64, nu=0.001, T=2.0, dt=0.002):
    crit = N // 3                      # 2/3-правило: храним моды |k| ≤ crit
    NP = int(1.5 * N)                  # 3/2-паддинг: перемножение без алиасинга
    k = np.fft.fftfreq(N, d=1.0 / N)   # целые волновые числа (период 2π)
    k0 = k[:, None]                    # производная по оси 0 (полная ось)
    k1 = k[None, : N // 2 + 1]         # производная по оси 1 (rfft, половинная)
    K2 = k0**2 + k1**2
    K2[0, 0] = 1.0
    kP = np.fft.fftfreq(NP, d=1.0 / NP)
    kP0 = kP[:, None]; kP1 = kP[None, : NP // 2 + 1]
    x = np.arange(N) / N * 2 * math.pi
    X, Y = np.meshgrid(x, x, indexing="ij")

    def div_free(fh):
        p = (k0 * fh[0] + k1 * fh[1]) / K2
        return np.stack([fh[0] - k0 * p, fh[1] - k1 * p])

    def trunc(fh):
        m = (np.abs(k0) <= crit) & (np.abs(k1) <= crit)
        return fh * m[None]

    def pad(fh):
        out = np.zeros((2, NP, NP // 2 + 1), complex)
        out[:, :crit+1, :crit+1] = fh[:, :crit+1, :crit+1]
        out[:, -crit:, :crit+1] = fh[:, -crit:, :crit+1]
        out[:, :crit+1, -crit:] = fh[:, :crit+1, -crit:]
        out[:, -crit:, -crit:] = fh[:, -crit:, -crit:]
        return out

    def unpad(nlp):
        out = np.zeros((2, N, N // 2 + 1), complex)
        out[:, :crit+1, :crit+1] = nlp[:, :crit+1, :crit+1]
        out[:, -crit:, :crit+1] = nlp[:, -crit:, :crit+1]
        out[:, :crit+1, -crit:] = nlp[:, :crit+1, -crit:]
        out[:, -crit:, -crit:] = nlp[:, -crit:, -crit:]
        return out

    def curl_h(fh):
        """ω̂ = i k0·v̂ − i k1·û  (ω = ∂₀v − ∂₁u)."""
        return 1j * k0 * fh[1] - 1j * k1 * fh[0]

    def div_h(fh):
        return 1j * k0 * fh[0] + 1j * k1 * fh[1]

    # Тейлор–Грин 2D
    u0 = np.sin(X) * np.cos(Y)
    v0 = -np.cos(X) * np.sin(Y)
    w = np.stack([u0, v0])
    wh = div_free(trunc(np.fft.rfftn(w)))
    om0 = np.fft.irfftn(curl_h(wh), s=(N, N))

    # --- (i): тождество ω' = cos θ_b·ω на 200 случайных div-free полях
    rng = np.random.default_rng(7)
    max_id = 0.0
    for _ in range(200):
        a = rng.normal(size=(N, N)); b2 = rng.normal(size=(N, N))
        fh = div_free(np.fft.rfftn(np.stack([a, b2])))
        om = np.fft.irfftn(curl_h(fh), s=(N, N))
        rot = np.stack([COS * fh[0] - SIN * fh[1], SIN * fh[0] + COS * fh[1]])
        om_r = np.fft.irfftn(curl_h(rot), s=(N, N))
        max_id = max(max_id, float(np.abs(om_r - COS * om).max()))

    # --- (ii): энергия при повороте сохраняется точно (в физическом пространстве)
    E = float((w**2).sum())
    w_rot = np.stack([COS * w[0] - SIN * w[1], SIN * w[0] + COS * w[1]])
    E_rot = float((w_rot**2).sum())
    energy_err = abs(E_rot - E) / E

    # --- (iii): div u' = −b·ω на поле Тейлора–Грина
    rot_tg = np.stack([COS * wh[0] - SIN * wh[1], SIN * wh[0] + COS * wh[1]])
    div_rot = np.fft.irfftn(div_h(rot_tg), s=(N, N))
    div_err = float(np.abs(div_rot - (-SIN * om0)).max())

    # --- NSE 2D: энергетическое тождество + max-принцип + непрерывный поворот
    def rhs(wh):
        # нелинейный член считаем на 3/2-сетке — без алиасинга (2/3-правило)
        fhp = pad(wh)
        uu = np.fft.irfftn(fhp, s=(NP, NP))
        d0 = np.fft.irfftn(1j * kP0 * fhp, s=(NP, NP))
        d1 = np.fft.irfftn(1j * kP1 * fhp, s=(NP, NP))
        adv = np.stack([uu[0] * d0[0] + uu[1] * d1[0],
                        uu[0] * d0[1] + uu[1] * d1[1]])
        return div_free(-unpad(np.fft.rfftn(adv))) - nu * K2 * wh

    n_steps = int(round(T / dt))
    E_prev = float((w[0]**2 + w[1]**2).sum() / 2 / N**2)
    diss_int = 0.0
    Z_prev = None
    om_max0 = float(np.abs(om0).max())
    om_max_viol = 0.0
    phi = dt * THETA
    cph, sph = math.cos(phi), math.sin(phi)
    max_dE_rot = 0.0
    wh_run = wh.copy()
    E_series = []

    for n in range(n_steps):
        k1a = rhs(wh_run)
        k2a = rhs(wh_run + 0.5 * dt * k1a)
        k3a = rhs(wh_run + 0.5 * dt * k2a)
        k4a = rhs(wh_run + dt * k3a)
        wh_run = wh_run + (dt / 6) * (k1a + 2 * k2a + 2 * k3a + k4a)

        om_now = np.fft.irfftn(curl_h(wh_run), s=(N, N))
        om_max_viol = max(om_max_viol, float(np.abs(om_now).max()) - om_max0)
        # энстрофия Z = ½‖ω‖²: трапецией, чтобы ∫2νZ dt был 2-го порядка
        Z_now = 0.5 * float((om_now**2).sum()) / N**2
        if Z_prev is not None:
            diss_int += nu * dt * (Z_prev + Z_now)  # 2ν·½(Zprev+Znow)dt = νdt(Zprev+Znow)
        Z_prev = Z_now

        if (n + 1) % 20 == 0:
            uu = np.fft.irfftn(wh_run, s=(N, N))
            E_series.append(float((uu**2).sum() / 2 / N**2))

        # непрерывный b-поворот
        uu = np.fft.irfftn(wh_run, s=(N, N))
        E_pre = float((uu**2).sum() / 2 / N**2)
        wh_rot = div_free(np.stack([cph * wh_run[0] - sph * wh_run[1],
                                    sph * wh_run[0] + cph * wh_run[1]]))
        uu2 = np.fft.irfftn(wh_rot, s=(N, N))
        E_post = float((uu2**2).sum() / 2 / N**2)
        max_dE_rot = max(max_dE_rot, abs(E_post - E_pre))
        wh_run = wh_rot

    E_final = E_series[-1] if E_series else E_prev
    energy_balance = abs(E_final - E_prev - (-1.0) * (diss_int))  # E(T) ≈ E(0) − ∫2νZ dt

    out = {
        "level": "L4",
        "params": {"N": N, "nu": nu, "T": T, "dt": dt, "theta_b_deg": float_theta_b_deg(),
                   "protocol": "continuous: per-step angle dt·θ_b; 2/3-правило через 3/2-паддинг"},
        "checks": {
            "тождество ω' = cos θ_b·ω, макс. невязка (200 полей)": max_id,
            "энергия при повороте: |E'−E|/E": energy_err,
            "div u' = −b·ω, макс. невязка": div_err,
            "максимум-принцип ω (превышение)": om_max_viol,
            "|ΔE| за один поворот (инжекции нет)": max_dE_rot,
            "энергетический баланс E(T)−E(0)+∫2νZ dt": energy_balance,
        },
        "ok": bool(max_id < 1e-12 and energy_err < 1e-13 and div_err < 1e-10
                   and om_max_viol < 1e-9 and max_dE_rot < 1e-7 and energy_balance < 1e-3),
    }
    RESULTS_DIR.mkdir(exist_ok=True)
    (RESULTS_DIR / "l4_nse_2d.json").write_text(
        json.dumps(out, ensure_ascii=False, indent=1), encoding="utf-8")
    print(f"L4: {'PASS' if out['ok'] else 'FAIL'} | ω'=cosθ·ω: {max_id:.2e}, "
          f"|ΔE| поворот: {max_dE_rot:.2e}, баланс: {energy_balance:.2e}")
    return out["ok"]


if __name__ == "__main__":
    sys.exit(0 if run() else 1)
